src/redux/actions/planActions.js

```diff
@@ -0,0 +1,329 @@
+    type: GET_PLAN_INFO_FAILURE,
+    lastUpdated: Date.now(),
+    isFetching: false,
+    isSynced: false
+  };
+}
+
+function getPlanInfoSuccess(plans) {
+  return {
+    type: GET_PLAN_INFO_SUCCESS,
+    lastUpdated: Date.now(),
+    isFetching: false,
+    isSynced: true,
+    plans: plans
+  };
+}
+
+/**************************** Logical functions ****************************/
+
+/**
+ * Performs an http POST get plan info request to server.
+ * Dispatches getUserInfoRequest to indicate the beginning of a getPlanInfo process.
+ * Dispatches getPlanInfoFailure to indicate the end of a failed getPlanInfo process.
+ * Dispatches getPlanSuccess to indicate the end of a successful getPlanInfo process.
+ * If getPlanInfo process succeeds, a plan list object is received and passed into
+ * getPlanInfoSuccess to be stored into state.
+ * @returns {function(*): Promise<Response | never>} dispatch results.
+ */
+function getPlanInfo(e_id) {
+  return (dispatch, getState) => {
+    dispatch(getPlanInfoRequest());
+    return fetch('http://localhost:4201/plan/get', {
+      headers: {
+        'Content-Type': 'application/json',
+        'x-access-token': localStorage.getItem('token')
+      },
+      method: 'post',
+      body: JSON.stringify({ e_id: e_id })
+    })
+      .then(response => response.json())
+      .then(json => {
+        if (json.length === 0) dispatch(getPlanInfoFailure());
+        else {
+          // Push newly added plan to state plan list
+          let plans = getState().masterReducer.plans.plans;
+          let trip_id = json.result[0].trip_id;
+          let e_id = json.result[0].e_id;
+          if(plans[trip_id] === undefined) plans[trip_id] = {};
+          plans[trip_id][e_id] = json.result[0];
+          dispatch(getPlanInfoSuccess(plans));
+        }
+      });
+  };
+}
+
+/**
+ * Performs an http GET plans request to server.
+ * Dispatches getPlans to indicate the beginning of a getPlans process.
+ * Dispatches getPlansFailure to indicate the end of a failed getPlans process.
+ * Dispatches getPlansSuccess to indicate the end of a successful getPlans process.
+ * If getPlans process succeeds, a plans object is received and passed into
+ * getPlansSuccess to be stored into state.
+ * @returns {function(*): Promise<Response | never>} dispatch results.
+ */
+export function getPlans() {
+  return dispatch => {
+    dispatch(getPlansRequest()); // Get plans request process has begun...
+    return fetch('http://localhost:4201/plan/get', {
+      headers: {
+        'Content-Type': 'application/json',
+        'x-access-token': localStorage.getItem('token')
+      },
+      method: 'GET'
+    })
+      .then(response => response.json())
+      .then(json => {
+        if (json.success === false) dispatch(getPlansFailure());
+        else {
+          let plans = [];
+          json.result.forEach(t => {
+            if (plans[t.trip_id] === undefined) plans[t.trip_id] = {};
+            plans[t.trip_id][t.e_id] = t;
```

Comment:
I like this 2D array architecture, allowing lookup by trip and getting all associated plans extremely quickly!

Reply:
Exactly. :)